[DNM] Mercury rising

ju-Skinner opened the pull request.

monkeypox8 commented on app/assets/javascripts/sage/system/inputhelper.js:
```diff
@@ -0,0 +1,149 @@
+      meter.value = zxcvbn(inputValue).score;
+      Sage.meter.updateMeter(meterClass);
+    } else {
+      meter.closest(".sage-meter").style.opacity = "0";
```
Open to suggestions on a better way to treat this - currently I'm visibly hiding the meter when the external lib isn't loaded, but `display: none` may be better suited.

AndrwM replied:
This seems good to me, could you make a comment here describing that this is a fallback in the case `zxcvbn` isn't loaded?

We'll have a better way to manage these dependancies once we're webpackin'.

AndrwM replied:
Another note, in previewing this I noticed that the meter is given 100% success even when a symbol isn't included:

![image](https://user-images.githubusercontent.com/565743/82494373-14f86980-9aaf-11ea-9d38-27771d1eb66d.png)

monkeypox8 replied:
Hmm, excellent point. Because the two items aren't coupled, I may have to add some sort of negative weighting value to the strength meter when all requirements haven't been fulfilled. The strength meter is meant more as a "guide" but ends up giving a false indication of a passing grade.

monkeypox8 commented on app/assets/javascripts/sage/system/inputhelper.js:
```diff
@@ -0,0 +1,149 @@
+  // ==================================================
+  // Variables
+  // ==================================================
+  var spcChar = /(?=.*[~`!@#$%^&*|(){}/=:;.,<>+-])/g;
```
Used https://regexr.com and https://regex101.com/ for testing these, but could use 👀 on validity of the strings here.

Will also want to refactor these at some point to allow for varying the minimum number or required symbols/numbers/characters

AndrwM replied:
Looks great, nice! I could see these becoming part of a global JS variables file later.

monkeypox8 commented on app/assets/javascripts/sage/system/inputhelper.js:
```diff
@@ -0,0 +1,149 @@
+          return {
+            id: ele.id,
+            type: ele.getAttribute("data-js-hint-type") || null,
+            minValue: ele.getAttribute("data-js-hint-min") || 0
```
Can remove this. For future adjustment of the minimum value required.

AndrwM commented on app/assets/stylesheets/sage/system/patterns/elements/_meter.scss:
```diff
@@ -5,7 +5,9 @@
 
 $-sage-meter-height: rem(18px);
-$-sage-meter-color-bg: sage-color(primary, 100);
+$-sage-meter-color-bg: sage-color(white);
+$-sage-meter-color-label: sage-color(charcoal, 100);
+$-sage-meter-label-shadow: 1px 1px 0 rgba(255, 255, 255, 0.67);
```
@philschanely – I agree about the `.sage-meter__label` contrast.

This is black at ~70%ish opacity via `rgba(0,0,0,0.7)`.
Maybe we could increase the contrast like this?

![screencast 2020-05-20 15-01-27](https://user-images.githubusercontent.com/565743/82491983-16c02e00-9aab-11ea-81f0-75d756e25371.gif)

monkeypox8 replied:
That does appear to be more readable. I may also lighten the tint of the background/meter color to increase the contrast some more.

AndrwM commented on app/assets/javascripts/sage/docs/inputhelper.js:
```diff
@@ -0,0 +1,20 @@
+Sage.docs.inputhelper = (function() {
+
+  // ==================================================
+  // Functions
+  // ==================================================
+
+  function init() {
+    var helperFieldset = document.querySelector('[data-js-example="input-helper"]');
+    var helperExample = helperFieldset.querySelector('.sage-input-helper');
+
+    helperExample.classList.add('sage-input-helper--visible');
+    helperExample.style.position = 'relative';
+  }
+
+
+  return {
+    init: init
+  };
+
+})();
```
This is used in the design system docs and it forces the `.sage-input-helper` to be open for the preview?

I wonder if we should namespace scripts like this so they're differentiated from the other docs-related scripting? Not sure if that's overthinking this, we'll likely have more items like this.

We could possibly do something like `Sage.docs.previewUtil.inputhelper` or `Sage.docs.example.inputhelper`. Though, I don't know much about the docs scripts or the common patterns happening there.

monkeypox8 replied:
Right, it's specifically for the preview. The docs scripts could use distinct filenames to help this as well, `inputhelper-example.js` or `inputhelper-docs.js` would clarify that it's not for public consumption. We don't have a lot of consistency as the docs structure has been evolving, so perhaps we'll need a follow-up issue cleaning up the documentation-only JS.

AndrwM commented on app/assets/javascripts/sage/system/inputhelper.js:
```diff
@@ -0,0 +1,148 @@
+    var metSym = false;
+    var metNum = false;
+
+    reqs.forEach(function(item) {
+      if (item.type === "characters") {
+        metChar = updateCriteria(pwLength(val), item.id);
+      } else if (item.type === "symbols") {
+        metSym = updateCriteria(pwSpecial(val), item.id);
+      } else if (item.type === "numbers") {
+        metNum = updateCriteria(pwNumber(val), item.id);
+      }
+    });
+
+    if (metChar && metSym && metNum) {
+      ele.parentElement.classList.remove(inputErrorClass);
+    }
+  }
+
+  // toggles item criteria
+  function updateCriteria(bool, itemID) {
+    var criteria = document.getElementById(itemID);
+
+    if (bool === true ) {
+      criteria.classList.add(passingClass);
+      return true;
+    } else {
+      criteria.classList.remove(passingClass);
+      return false;
+    }
+  }
```
Nice!

AndrwM commented on app/assets/javascripts/sage/system/inputhelper.js:
```diff
@@ -0,0 +1,148 @@
+
+    var meter = helper.querySelector(".sage-meter__bar");
+
+    if (Sage.util.isIE()) {
+      focusBlurIE(field);
+    }
```
😢 The things we do for you IE.

AndrwM commented on app/assets/stylesheets/sage/system/patterns/elements/_meter.scss:
```diff
@@ -5,7 +5,9 @@
+$-sage-meter-label-shadow: 1px 1px 0 rgba(255, 255, 255, 0.67);
 $-sage-meter-bar-low: sage-color(red);
 $-sage-meter-bar-med: sage-color(orange);
 $-sage-meter-bar-optimum: sage-color(sage, 200);
```
What do you think about using `$color-success` or `$color-green-light` for this?

monkeypox8 commented:
@AndrwM Thanks for the feedback! Made some edits to address the issues you noted with scoring along with @philschanely 's notes on the label contrast.

cmcfadzean commented on app/views/sage/examples/objects/input_helper/_props.html.erb:
```diff
@@ -0,0 +1,6 @@
+<tr>
```
Can probable delete the content here if not applicable at this time

cmcfadzean commented on app/views/sage/examples/objects/input_helper/_rules_dont.html.erb:
```diff
@@ -0,0 +1,3 @@
+<ul>
```
Can probable delete the content here if not applicable at this time

cmcfadzean commented on app/assets/stylesheets/sage/system/patterns/elements/_meter.scss:
```diff
@@ -5,11 +5,13 @@
-$-sage-meter-bar-max: sage-color(sage);
+$-sage-meter-color-bg: sage-color(white);
+$-sage-meter-color-label: rgba(sage-color(black), 0.7);
+$-sage-meter-label-shadow: 1px 1px 0 rgba(255, 255, 255, 0.67);
```
In the screenshots, the label shadow is throwing me off. It seems out of alignment with the other look and feel. 
Curious how design feels about this. I am tempted to say that the label is not even needed and it could just be the color indicator bar and that being clear enough.

monkeypox8 replied:
Good call, I'll check with design. We don't have a full comp for this yet, so the bar label is only necessary for a11y purposes.

monkeypox8 commented:
Feedback from design was to allow the label of the meter to also indicate strength (examples attached), also allowing us to reduce the height of the bar. To prevent this PR from languishing further, I'm planning to address the revisions in a follow-up.

|  Example 1   |  Example 2  |
|--------|--------|
|![image](https://user-images.githubusercontent.com/816579/83314884-64d2e100-a1d1-11ea-8867-9502ee44854e.png)|![image (1)](https://user-images.githubusercontent.com/816579/83314891-72886680-a1d1-11ea-8fc6-8f1488ba9e4b.png)|

Commit:
Tokens - Adds New Rebrand Colors and Font (#222)

* style: adds new rebrand color tokens

* style: remaps old charcoal semantic tokens to new gray

* style: builds new color tokens

* style: updates yellow 200

* style: updates token values

* style: builds updated token set

* style: updates tokens with new names

* style: updates semantic references

* style: updates Circular font family token to Greet

* style: corrects color token values

* style: fixes semantic token references

* style: build new token set

* style: changes "gray" to "gray"

* style: builds grey tokens

---------

Co-authored-by: Kajabi Automations Bot <120066875+dev-kjbot@users.noreply.github.com>

Commit:
Tokens: Adds Letter Spacing (#223)

* style: adds new letter spacing tokens and removes 0 from some font size and line height values

* style: builds new letter spacing tokens

* style: updates font size, removes typography semantic tokens, and removes leading zero on remaining tokens

* style: builds new tokens

---------

Co-authored-by: Kajabi Automations Bot <120066875+dev-kjbot@users.noreply.github.com>

Commit:
Tokens: Update Typography Tokens (#225)

* style: updates font weight values, adds a new line height value, and correct font family value

* style: builds new type tokens

* style: fixes type changes

* style: builds fixed tokens

---------

Co-authored-by: Kajabi Automations Bot <120066875+dev-kjbot@users.noreply.github.com>

Commit:
Tokens: Adds Initial Semantic Tokens (#226)

* style: adds initial semantic tokens

* style: builds new semantic tokens

---------

Co-authored-by: Kajabi Automations Bot <120066875+dev-kjbot@users.noreply.github.com>

Commit:
Tokens: Creates Semantic Type Tokens (#227)

* style: builds new semantic type tokens

* style: readds missing tokens in JSON

* style: builds missing tokens

---------

Co-authored-by: Kajabi Automations Bot <120066875+dev-kjbot@users.noreply.github.com>

kajabi-bot committed:
chore: publish

 - @pine-ds/core@2.0.0-mercury.0
 - @pine-ds/doc-components@2.0.0-mercury.0
 - @pine-ds/react@2.0.0-mercury.0

kajabi-bot committed:
Publish

 - @pine-ds/core@3.0.0-mercury.0
 - @pine-ds/doc-components@3.0.0-mercury.0
 - @pine-ds/react@3.0.0-mercury.0

Commit:
styles: updates font weight, letter spacing, and font size (#230)

* chore: update doc-components version (#228)

* style: reverts semantic font weights back to 100 scale

* style: updates letter spacing for body type

* style: updates xs font size to 12px

* style: build updated tokens

* chore: package-lock update

* ci: update @tokens-studio/sd-transforms package to the latest version

* chore: adds necessary preprocessor to script file to fix shadow tokens

---------

Co-authored-by: Julian Skinner <ju-Skinner@users.noreply.github.com>
Co-authored-by: Kajabi Automations Bot <120066875+dev-kjbot@users.noreply.github.com>
Co-authored-by: Julian Skinner <ju.skinner@gmail.com>

kajabi-bot committed:
Publish

 - @pine-ds/core@4.0.0-mercury.0
 - @pine-ds/doc-components@4.0.0-mercury.0
 - @pine-ds/react@4.0.0-mercury.0

pixelflips committed:
style(link): correct token naming for icon spacing (#234)

kajabi-bot committed:
Publish

 - @pine-ds/core@5.0.0-mercury.0
 - @pine-ds/doc-components@5.0.0-mercury.0
 - @pine-ds/react@5.0.0-mercury.0

Commit:
Typography: add new fonts (#233)

* style: adds new fonts to Pine

* chore: fixes style linting issues

kajabi-bot committed:
Publish

 - @pine-ds/core@6.0.0-mercury.0
 - @pine-ds/doc-components@6.0.0-mercury.0
 - @pine-ds/react@6.0.0-mercury.0

ju-Skinner committed:
gi: update versions to correct version numbers

the builds for mercury were using premajor when it should have been preminor

after the first mercury release

kajabi-bot committed:
Publish

 - @pine-ds/core@2.1.0-mercury.0
 - @pine-ds/doc-components@2.1.0-mercury.0
 - @pine-ds/react@2.1.0-mercury.0

Commit:
Accordion: Minor Styling Updates (#238)

* style: fixes accordion body padding and some deprecated tokens

* style: updates font styles of summary trigger

kajabi-bot committed:
Publish

 - @pine-ds/core@2.2.0-mercury.0
 - @pine-ds/doc-components@2.2.0-mercury.0
 - @pine-ds/react@2.2.0-mercury.0

QuintonJason committed:
style: update various color tokens (#236)

* style: update various color tokens

* style: update various color tokens

kajabi-bot committed:
Publish

 - @pine-ds/core@2.3.0-mercury.0
 - @pine-ds/doc-components@2.3.0-mercury.0
 - @pine-ds/react@2.3.0-mercury.0

QuintonJason committed:
style: various token changes (#239)

* fix: update font size refs

* fix: remove leading zero and correct font-size

* fix: update charcoal to grey

* fix: add back zeroes

* fix: added zeroes to semantic tokens

* fix: added zeroes to component token refs

* fix: update accordion padding

* fix: added back 0

kajabi-bot committed:
Publish

 - @pine-ds/core@2.4.0-mercury.0
 - @pine-ds/doc-components@2.4.0-mercury.0
 - @pine-ds/react@2.4.0-mercury.0

QuintonJason committed:
fix: update box shadow tokens (#241)

* fix: add box shadow tokens

* fix: add box shadow build tokens

* fix: update box-shadow tokens

* fix: update box shadow tokens

* fix: add and apply box shadow tokens

* fix: update sort function

* fix: typo

kajabi-bot committed:
Publish

 - @pine-ds/core@2.5.0-mercury.0
 - @pine-ds/doc-components@2.5.0-mercury.0
 - @pine-ds/react@2.5.0-mercury.0

kajabi-bot committed:
Publish

 - @pine-ds/core@2.6.0-mercury.0
 - @pine-ds/doc-components@2.6.0-mercury.0
 - @pine-ds/react@2.6.0-mercury.0

Commit:
style: updates accordion styling to better match Figma specs (#245)

kajabi-bot committed:
Publish

 - @pine-ds/core@2.7.0-mercury.0
 - @pine-ds/doc-components@2.7.0-mercury.0
 - @pine-ds/react@2.7.0-mercury.0

kajabi-bot committed:
Publish

 - @pine-ds/core@2.8.0-mercury.0
 - @pine-ds/doc-components@2.8.0-mercury.0
 - @pine-ds/react@2.8.0-mercury.0

kajabi-bot committed:
Publish

 - @pine-ds/core@2.9.0-mercury.0
 - @pine-ds/doc-components@2.9.0-mercury.0
 - @pine-ds/react@2.9.0-mercury.0

kajabi-bot committed:
Publish

 - @pine-ds/core@2.10.0-mercury.0
 - @pine-ds/doc-components@2.10.0-mercury.0
 - @pine-ds/react@2.10.0-mercury.0

pixelflips committed:
style: update focus state width and color

ju-Skinner committed:
style(tabs): change the active tab color to mercury 500 (#248)

kajabi-bot committed:
Publish

 - @pine-ds/core@2.11.0-mercury.0
 - @pine-ds/doc-components@2.11.0-mercury.0
 - @pine-ds/react@2.11.0-mercury.0

pixelflips committed:
style: correct lint error

pixelflips committed:
style: correct lint error

pixelflips committed:
style: remove extra semicolon

kajabi-bot committed:
Publish

 - @pine-ds/core@2.12.0-mercury.0
 - @pine-ds/doc-components@2.12.0-mercury.0
 - @pine-ds/react@2.12.0-mercury.0

ju-Skinner committed:
style(button): change border radius to 9999px (#251)

ju-Skinner committed:
style(button): change the colors for the accent variant to be purple (#253)

ju-Skinner committed:
ci: fix lerna version command to use preminor

kajabi-bot committed:
Publish

 - @pine-ds/core@2.15.0-mercury.0
 - @pine-ds/doc-components@2.15.0-mercury.0
 - @pine-ds/react@2.15.0-mercury.0

pixelflips committed:
style: add relative position for focus (#259)

kajabi-bot committed:
Publish

 - @pine-ds/core@2.16.0-mercury.0
 - @pine-ds/doc-components@2.16.0-mercury.0
 - @pine-ds/react@2.16.0-mercury.0

Commit:
chore: fixes issue with sd-transforms caused by version update (#260)

kajabi-bot committed:
Publish

 - @pine-ds/core@2.17.0-mercury.0
 - @pine-ds/doc-components@2.17.0-mercury.0
 - @pine-ds/react@2.17.0-mercury.0

Commit:
style: updates button font family and font weight (#258)
